File: CS/WinForms.Client/RemoteContextUtils.cs
using System;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using DataModel.Shared.BusinessObjects;
using DevExpress.EntityFrameworkCore.Security.MiddleTier.ClientServer;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Security;
using DevExpress.ExpressApp.Security.ClientServer;
using Microsoft.EntityFrameworkCore;

namespace WinForms.Client {
    public class RemoteContextUtils {
        public static DbContextOptions<DXApplication1EFCoreDbContext> Options { get; set; }

        public static WebApiSecuredDataServerClient SecuredDataServerClient { get; set; }

        public static void Logoff() {
            Options = null;
            SecuredDataServerClient = null;
        }

        public static bool IsLogin {
            get {
                return Options != null;
            }
        }

        public static bool IsGranded(Type type, string operation) {
            var request = new SerializablePermissionRequest(type, null, null, operation);
            return ((IMiddleTierServerSecurity)SecuredDataServerClient).IsGranted(request);
        }

        public static DXApplication1EFCoreDbContext GetDBContext() {
            return new DXApplication1EFCoreDbContext(Options);
        }

        public static DbContextOptions<DXApplication1EFCoreDbContext> CreateDbContextOptions(WebApiSecuredDataServerClient securedClient) {
            var builder = new DbContextOptionsBuilder<DXApplication1EFCoreDbContext>();
            return builder
                .UseLazyLoadingProxies()
                .UseChangeTrackingProxies()
                .UseMiddleTier(securedClient)
                .Options as DbContextOptions<DXApplication1EFCoreDbContext>;
        }

        public static WebApiSecuredDataServerClient CreateSecuredClient(string url, string login, string password) {
            // Connect to an EF Core Middle Tier Security Application from Non-XAF Applications
            // https://docs.devexpress.com/eXpressAppFramework/404398/data-security-and-safety/security-system/non-xaf/connect-to-the-efcore-middle-tier-service-from-non-xaf-applications

            var httpClient = new HttpClient();
            httpClient.BaseAddress = new Uri(url);
            httpClient.DefaultRequestHeaders.Add("Accept", "application/json");

            var securedClient = new WebSocketSecuredDataServerClient(httpClient, XafTypesInfo.Instance);
            securedClient.CustomAuthenticate += (sender, arguments) => {
                arguments.Handled = true;
                HttpResponseMessage msg = arguments.HttpClient.PostAsJsonAsync("api/Authentication/Authenticate", (AuthenticationStandardLogonParameters)arguments.LogonParameters).GetAwaiter().GetResult();
                string token = (string)msg.Content.ReadFromJsonAsync(typeof(string)).GetAwaiter().GetResult();
                if(msg.StatusCode == HttpStatusCode.Unauthorized) {
                    throw new UserFriendlyException(token);
                }
                msg.EnsureSuccessStatusCode();
                arguments.HttpClient.DefaultRequestHeaders.Authorization
                    = new AuthenticationHeaderValue("bearer", token);
            };

            securedClient.Authenticate(new AuthenticationStandardLogonParameters(login, password));
            ((IMiddleTierServerSecurity)securedClient).Logon();

            return securedClient;
        }
    }
}

File: VB/WinForms.Client/RemoteContextUtils.vb
Imports System
Imports System.Net
Imports System.Net.Http
Imports System.Net.Http.Headers
Imports System.Net.Http.Json
Imports DataModel.Shared.BusinessObjects
Imports DataModel.Shared.DataModel.Shared.BusinessObjects
Imports DevExpress.EntityFrameworkCore.Security.MiddleTier.ClientServer
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Security
Imports DevExpress.ExpressApp.Security.ClientServer
Imports Microsoft.EntityFrameworkCore

Namespace WinForms.Client

    Public Class RemoteContextUtils

        Public Shared Property Options As DbContextOptions(Of DXApplication1EFCoreDbContext)

        Public Shared Property SecuredDataServerClient As WebApiSecuredDataServerClient

        Public Shared Sub Logoff()
            Options = Nothing
            SecuredDataServerClient = Nothing
        End Sub

        Public Shared ReadOnly Property IsLogin As Boolean
            Get
                Return Options IsNot Nothing
            End Get
        End Property

        Public Shared Function IsGranded(ByVal type As Type, ByVal operation As String) As Boolean
            Dim request = New SerializablePermissionRequest(type, Nothing, Nothing, operation)
            Return CType(SecuredDataServerClient, IMiddleTierServerSecurity).IsGranted(request)
        End Function

        Public Shared Function GetDBContext() As DXApplication1EFCoreDbContext
            Return New DXApplication1EFCoreDbContext(Options)
        End Function

        Public Shared Function CreateDbContextOptions(ByVal securedClient As WebApiSecuredDataServerClient) As DbContextOptions(Of DXApplication1EFCoreDbContext)
            Dim builder = New DbContextOptionsBuilder(Of DXApplication1EFCoreDbContext)()
            Return TryCast(builder.UseLazyLoadingProxies().UseChangeTrackingProxies().UseMiddleTier(securedClient).Options, DbContextOptions(Of DXApplication1EFCoreDbContext))
        End Function

        Public Shared Function CreateSecuredClient(ByVal url As String, ByVal login As String, ByVal password As String) As WebApiSecuredDataServerClient
            ' Connect to an EF Core Middle Tier Security Application from Non-XAF Applications
            ' https://docs.devexpress.com/eXpressAppFramework/404398/data-security-and-safety/security-system/non-xaf/connect-to-the-efcore-middle-tier-service-from-non-xaf-applications
            Dim httpClient = New HttpClient()
            httpClient.BaseAddress = New Uri(url)
            httpClient.DefaultRequestHeaders.Add("Accept", "application/json")
            Dim securedClient = New WebSocketSecuredDataServerClient(httpClient, XafTypesInfo.Instance)
            AddHandler securedClient.CustomAuthenticate, Sub(sender, arguments)
                                                             arguments.Handled = True
                                                             Dim msg As HttpResponseMessage = arguments.HttpClient.PostAsJsonAsync("api/Authentication/Authenticate", CType(arguments.LogonParameters, AuthenticationStandardLogonParameters)).GetAwaiter().GetResult()
                                                             Dim token As String = CStr(msg.Content.ReadFromJsonAsync(GetType(String)).GetAwaiter().GetResult())
                                                             If msg.StatusCode = HttpStatusCode.Unauthorized Then
                                                                 Throw New UserFriendlyException(token)
                                                             End If

                                                             msg.EnsureSuccessStatusCode()
                                                             arguments.HttpClient.DefaultRequestHeaders.Authorization = New AuthenticationHeaderValue("bearer", token)
                                                         End Sub
            securedClient.Authenticate(New AuthenticationStandardLogonParameters(login, password))
            CType(securedClient, IMiddleTierServerSecurity).Logon()
            Return securedClient
        End Function
    End Class
End Namespace
